examples/Spec/Rational.hs

```diff
@@ -0,0 +1,192 @@
+      :&&: (Var (Num IsSmall) :->: Var RatSmall)
+      :&&: Not (Var $ Den IsZero)
+      :&&: (Var (Num IsZero) :<->: Var RatZero)
+      :&&: ExactlyOne [Var RatPos, Var RatZero, Var RatNeg]
+      :&&: (Var RatZero :||: (ExactlyOne [Var (Num IsNegative), Var (Den IsNegative)] :<->: Var RatNeg))
+      :&&: (((Var (Num IsMaxBound) :||: Var (Num IsMinBound)) :&&: Var RatSmall) :->: Var (Den IsLarge))
+      :&&: ((Var (Den IsMaxBound) :||: Var (Den IsMinBound)) :->: Var RatSmall)
+
+instance HasLogicalModel RatProp Rat where
+  satisfiesProperty RatPos r = asRational r > 0
+  satisfiesProperty RatZero r = asRational r == 0
+  satisfiesProperty RatNeg r = asRational r < 0
+  satisfiesProperty RatLarge r = abs (asRational r) > 10
+  satisfiesProperty RatSmall r = abs (asRational r) <= 10
+  satisfiesProperty (Num p) r = satisfiesProperty p (num r)
+  satisfiesProperty (Den p) r = satisfiesProperty p (den r)
+
+instance HasAbstractions RatProp Rat where
+  abstractions =
+    [ WrapAbs $
+        ProductAbstraction
+          { abstractionName = "numerator"
+          , propertyAbstraction = abstractsProperties Num
+          , productModelAbstraction = lens num (\r n -> r{num = n})
+          }
+    , WrapAbs $
+        ProductAbstraction
+          { abstractionName = "denominator"
+          , propertyAbstraction = abstractsProperties Den
+          , productModelAbstraction = lens den (\r d -> r{den = d})
+          }
+    ]
+
+instance HasPermutationGenerator RatProp Rat where
+  -- Some of the morphisms generated are nonsensicle ie. make Large >>> fix sign >>> fix small
+  -- so we set this to true to disable the check that each morphism is usefull
+  allowRedundentMorphisms _ = True
+
+  generators =
+    ( abstractionMorphisms
+        ++ [ Morphism
+              { name = "make Large"
+              , match = Not $ Var RatLarge
+              , contract = remove RatSmall >> add RatLarge
+              , morphism = pure
+              }
+           , Morphism
+              { name = "make Small"
+              , match = Not $ Var RatSmall
+              , contract = remove RatLarge >> add RatSmall
+              , morphism = pure
+              }
+           ]
+    )
+      -- The relationship between RatNet, RatPos and RatZero and the sub model properties
+      -- is simple and fully described by the model, so it's possible to fix the invalid morphisms like numerator of negate
+      -- by simply branching to any rational sign and failing on the branches where the logic is not satisfied
+      >>> [ Morphism
+              { name = "fix sign"
+              , match = Yes
+              , contract =
+                  removeAll [RatZero, RatPos, RatNeg]
+                    >> branches
+                      [ add RatZero >> matches logic
+                      , add RatPos >> matches logic
+                      , add RatNeg >> matches logic
+                      ]
+              , morphism = pure
+              }
+          ]
```

Comment:
This is a really interesting fix! My understanding is this will generate (along with the fixes) some morphisms that can never do anything since they don't make sense under the model. The nonsensical fixes will be ignored when we construct the DAG however so this is fine! :)